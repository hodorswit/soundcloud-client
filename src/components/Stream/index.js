import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { actionCreators as authActionCreators } from "../../ducks/auth";
import { actionCreators as trackActionCreators } from "../../ducks/track";
import Stream from "./presenter";

function mapStateToProps(state) {
  const { user } = state.auth;
  const { trackIds, trackEntities, activeTrackId } = state.track;

  return {
    user,
    trackIds,
    trackEntities,
    activeTrackId
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onAuth: bindActionCreators(authActionCreators.doAuth, dispatch),
    onPlay: bindActionCreators(trackActionCreators.doPlayTrack, dispatch),
    onLike: bindActionCreators(trackActionCreators.doLikeTrack, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Stream);
