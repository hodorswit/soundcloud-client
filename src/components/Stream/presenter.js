import React, { Component } from "react";
import ReactDOM from "react-dom";
import { observer } from "mobx-react";

class LikeButton extends Component {
  render() {
    const { track } = this.props;
    return (
      <span>
        {track.user_favorite
          ? <button type="button" onClick={() => this.likeTrack(track)}>
              Unlike
            </button>
          : <button type="button" onClick={() => this.likeTrack(track)}>
              Like
            </button>}
      </span>
    );
  }

  likeTrack = track => {
    this.props.onLike(track);
  };
}

@observer class Stream extends Component {
  componentDidUpdate() {
    const audioElement = ReactDOM.findDOMNode(this.refs.audio);

    if (!audioElement) return;

    const { activeTrackId } = this.props;

    if (activeTrackId) {
      audioElement.play();
    } else {
      audioElement.pause();
    }
  }

  render() {
    const {
      me,
      tracks,
      activeTrack,
      clientId,
      onAuth,
      onPlay,
      onLike
    } = this.props;

    return (
      <div>
        <div>
          {me
            ? <div>{me.username}</div>
            : <button onClick={onAuth} type="button">Login</button>}
        </div>
        <br />
        <div>
          {tracks.map((track, key) => {
            return (
              <div className="track" key={key}>
                {track.title}
                <button type="button" onClick={() => onPlay(track)}>
                  Play
                </button>
                <LikeButton track={track} onLike={onLike} />
              </div>
            );
          })}
        </div>
        {activeTrack
          ? <div>
              <div>
                Playing: {activeTrack.title}
                <LikeButton track={activeTrack} onLike={onLike} />
              </div>
              <audio
                id="audio"
                ref="audio"
                src={`${activeTrack.stream_url}?client_id=${clientId}`}
              />
            </div>
          : null}
      </div>
    );
  }
}

export default Stream;
