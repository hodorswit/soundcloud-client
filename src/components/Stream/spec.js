import React from "react";
import Stream from "./presenter";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";

describe("Stream", function() {
  const props = {
    trackIds: ["x", "y"],
    trackEntities: { x: { title: "x" }, y: { title: "y" } }
  };

  it("snapshot", () => {
    const component = renderer.create(<Stream tracks={props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders two tracks", () => {
    const component = shallow(<Stream tracks={props} />);
    expect(component.find(".track")).toHaveLength(2);
  });
});
