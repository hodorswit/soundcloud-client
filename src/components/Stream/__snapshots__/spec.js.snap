exports[`Stream snapshot 1`] = `
<div>
  <div>
    <button
      onClick={undefined}
      type="button">
      Login
    </button>
  </div>
  <br />
  <div />
</div>
`;
